### ISSUE TITLE
[API对照表] 关于pytorch和paddle的scatter

### ISSUE THREAD

**Issue by geoyee** [PaddlePaddle/X2Paddle#724]
请问能否增加一个关于pytorch和paddle的scatter的对比和讲解，以及怎么用paddle的flatten、scatter_nd等API组合达到pytorch的scatter的效果

**Comment by wjj19950828** [comment]
好的，这边会排个期支持一下～

**Comment by wjj19950828** [comment]
@geoyee 您好，已添加torch.scatter api映射文档以及如何组合实现，相关PR：https://github.com/PaddlePaddle/X2Paddle/pull/726

**Comment by geoyee** [comment]
> @geoyee 您好，已添加torch.scatter api映射文档以及如何组合实现，相关PR：#726

谢谢，明天试试满足的话就close这个issue

**Comment by geoyee** [comment]
已解决我当前问题，只是dim==2的情况要稍微改下

**Comment by geoyee** [comment]
我参考了这个组合实现，但任务中需要的是`dim==2`的情况。于是我根据代码写了`dim==2`的情况，我创造了一个测试，它给出了相同的结果，所以开始我认为它是对的，如下：
``` python
# pytorch
src = torch.arange(1, 31).reshape((2, 5, 3))
tindex = torch.tensor([[[0, 1, 4], [1, 2, 3], [2, 3, 4]], [[2, 1, 3], [1, 4, 2], [2, 3, 1]]])
print(torch.zeros(3, 5, 5, dtype=src.dtype).scatter_(2, tindex, src))
# 输出
# tensor([[[ 1,  2,  0,  0,  3],
#          [ 0,  4,  5,  6,  0],
#          [ 0,  0,  7,  8,  9],
#          [ 0,  0,  0,  0,  0],
#          [ 0,  0,  0,  0,  0]],
#         [[ 0, 17, 16, 18,  0],
#          [ 0, 19, 21,  0, 20],
#          [ 0, 24, 22, 23,  0],
#          [ 0,  0,  0,  0,  0],
#          [ 0,  0,  0,  0,  0]],
#         [[ 0,  0,  0,  0,  0],
#          [ 0,  0,  0,  0,  0],
#          [ 0,  0,  0,  0,  0],
#          [ 0,  0,  0,  0,  0],
#          [ 0,  0,  0,  0,  0]]])

# paddle
x = paddle.zeros([3, 5, 5], dtype="int64")
updates = paddle.arange(1, 31).reshape([2, 5, 3])
index = paddle.to_tensor([[[0, 1, 4], [1, 2, 3], [2, 3, 4]], [[2, 1, 3], [1, 4, 2], [2, 3, 1]]]).astype("int64")
i, j, k = index.shape
grid_x , grid_y, grid_z = paddle.meshgrid(paddle.arange(i), paddle.arange(j), paddle.arange(k))
index = paddle.stack([grid_x.flatten(), grid_y.flatten(), index.flatten()], axis=1)
updates_index = paddle.stack([grid_x.flatten(), grid_y.flatten(), grid_z.flatten()], axis=1)
updates = paddle.gather_nd(updates, index=updates_index)
print(paddle.scatter_nd_add(x, index, updates))
# 输出
# Tensor(shape=[3, 5, 5], dtype=int64, place=CPUPlace, stop_gradient=True,
#        [[[1 , 2 , 0 , 0 , 3 ],
#          [0 , 4 , 5 , 6 , 0 ],
#          [0 , 0 , 7 , 8 , 9 ],
#          [0 , 0 , 0 , 0 , 0 ],
#          [0 , 0 , 0 , 0 , 0 ]],
#         [[0 , 17, 16, 18, 0 ],
#          [0 , 19, 21, 0 , 20],
#          [0 , 24, 22, 23, 0 ],
#          [0 , 0 , 0 , 0 , 0 ],
#          [0 , 0 , 0 , 0 , 0 ]],
#         [[0 , 0 , 0 , 0 , 0 ],
#          [0 , 0 , 0 , 0 , 0 ],
#          [0 , 0 , 0 , 0 , 0 ],
#          [0 , 0 , 0 , 0 , 0 ],
#          [0 , 0 , 0 , 0 , 0 ]]])
```
但在网络中一个较复杂的情况下却不相同，我不知道知道是`dim==2`不应该这样写吗还是什么原因，我发现一些位置是对的，我认为可能是`paddle.scatter_nd_add`的问题，可能有同位置的index。如下：
``` python
def torch_scatter(x, dim, index, src):
    i, j, k = index.shape
    index = index.astype("int64")
    grid_x, grid_y, grid_z = paddle.meshgrid(paddle.arange(i), paddle.arange(j), paddle.arange(k))
    if dim == 0:
        index = paddle.stack([index.flatten(), grid_y.flatten(), grid_z.flatten()], axis=1)
    elif dim == 1:
        index = paddle.stack([grid_x.flatten(), index.flatten(), grid_z.flatten()], axis=1)
    else:  # dim == 2
        index = paddle.stack([grid_x.flatten(), grid_y.flatten(), index.flatten()], axis=1)
    src_index = paddle.stack([grid_x.flatten(), grid_y.flatten(), grid_z.flatten()], axis=1)
    src = paddle.gather_nd(src, index=src_index)
    paddle_out = paddle.scatter_nd_add(x, index, src)
    return paddle_out

x_low = np.load("x_low.npy")
low_edge_indices = np.load("low_edge_indices.npy").astype("int64")
fusion_edge_feat = np.load("fusion_edge_feat.npy")
print(x_low.shape)
print(low_edge_indices.shape)
print(fusion_edge_feat.shape)

tx = torch.Tensor(x_low)
te = torch.Tensor(low_edge_indices).to(torch.int64)
tf = torch.Tensor(fusion_edge_feat)
ty = tx.scatter(2, te, tf)
print("pytorch:")
print(ty[0, :5, :5].detach().numpy())

px = paddle.to_tensor(x_low)
pe = paddle.to_tensor(low_edge_indices, dtype="int64")
pf = paddle.to_tensor(fusion_edge_feat)
py = torch_scatter(px, 2, pe, pf)
print("paddle:")
print(py[0, :5, :5].numpy())

# 输出
# (2, 256, 16384)
# (2, 256, 128)
# (2, 256, 128)
# pytorch:
# [[2.8692201e-01 0.0000000e+00 7.2697005e-16 0.0000000e+00 2.2112370e-02]
#  [1.1013821e+01 2.0849142e+00 9.6707897e+00 1.3716705e+00 9.6912394e+00]
#  [8.2276529e-01 0.0000000e+00 8.2276517e-01 0.0000000e+00 8.2276517e-01]
#  [0.0000000e+00 0.0000000e+00 0.0000000e+00 0.0000000e+00 0.0000000e+00]
#  [3.4086432e+00 1.0611908e+00 3.3762336e+00 1.1810933e+00 3.3749661e+00]]
# paddle:
# [[1.3162262e+00 0.0000000e+00 7.2697005e-16 0.0000000e+00 7.0481062e-02]
#  [1.4029447e+01 2.0849142e+00 1.1031783e+01 1.3716705e+00 1.1130833e+01]
#  [8.2276529e-01 0.0000000e+00 8.2276517e-01 0.0000000e+00 8.2276517e-01]
#  [0.0000000e+00 0.0000000e+00 0.0000000e+00 0.0000000e+00 0.0000000e+00]
#  [4.7450027e+00 1.0611908e+00 4.6704397e+00 1.1810933e+00 4.5953026e+00]]
```
麻烦帮忙看看，怎样才能和pytorch的这个scatter对齐，这个npy的文件我放在这儿了：[Link](https://next.a-boat.cn:2021/s/zewBR83RgAi2r2e)，感谢~

**Comment by wjj19950828** [comment]
好的，这边看一下～

**Comment by geoyee** [comment]
> 好的，这边看一下～

请问有结果了吗
